### PR TITLE
Merge SRA Identity & Auth to master

### PR DESCRIPTION
**Pull request by gosar**
## Motivation and Context

Refactor Identity & Auth components in the SDKs to follow the Smithy Reference Architecture (SRA). This is a backwards compatible refactor, that standardizes modular interfaces and improves the customizability for the SDKs.

## Modifications

Lots of changes! Broadly,
* New modules with new interfaces for Identity & Auth types: IdentityProvider, HttpSigner, AuthScheme, etc. And corresponding implementation modules.
* Introduce new auth scheme resolution approach, which gives control on how to choose the auth scheme and configure the signing/identity resolution process.
* Codegen changes to configure the clients to use the new auth/identity resolution logic.
* Update core to support new auth resolution and execution using the new interfaces.
* Update the client/request interfaces to allow customer to configure the client using these new interfaces, to support different customization use cases.
* Introduce a service-level codegen flag to allow service-by-service migration to the new SRA approach.

## Testing

Range of testing: unit, functional, integ, backwards compatibility, etc.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
        (Relying on CI)
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/AsyncSignerOverrideTest.java [hunk 1]
```diff
@@ -57,4 +58,10 @@ public void test_signerOverriddenForStreamingInput_takesPrecedence() {
 
         verify(mockSigner).sign(any(SdkHttpFullRequest.class), any(ExecutionAttributes.class));
     }
+
+    // TODO(sra-identity-and-auth): Add test for SRA way of overriding signer to assert that overridden signer is used.
+    //  To do this, need ability to inject AuthScheme which uses mock HttpSigner. This is pending https://i.amazon.com/SMITHY-1450
```
Comment (dagnir) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/AsyncSignerOverrideTest.java:63]: removethe URL

**Comment by dagnir** [comment]
Are we tracking all of the TODOs introduced in this change?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-bearer-auth-client-builder-interface.java
[the remark this answers is not in the document]
Comment (gosar) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/builder/test-bearer-auth-client-builder-interface.java:1]: Being fixed in https://github.com/aws/aws-sdk-java-v2/pull/4581

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/SigningStage.java [hunk 2]
```diff
@@ -52,57 +66,115 @@ public SigningStage(HttpClientDependencies dependencies) {
...
-            if (signer instanceof AsyncRequestBodySigner) {
-                //Transform request body provider with signing operator
-                AsyncRequestBody transformedRequestProvider =
-                    ((AsyncRequestBodySigner) signer)
-                        .signAsyncRequestBody(signedRequest, context.requestProvider(), context.executionAttributes());
-                context.requestProvider(transformedRequestProvider);
-            }
-            updateInterceptorContext(signedRequest, context.executionContext());
-            return signedRequest;
+        // TODO: This case does not apply to SigningStage as event stream operations are not supported by SyncClients that
```
[the remark this answers is not in the document]
Comment (gosar) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/SigningStage.java:150]: Removing in https://github.com/aws/aws-sdk-java-v2/pull/4574

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/AsyncSignerOverrideTest.java (hunk 1 above)
Comment (gosar) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/AsyncSignerOverrideTest.java:63]: Removing in https://github.com/aws/aws-sdk-java-v2/pull/4574

**Comment by gosar** [comment]
> Are we tracking all of the TODOs introduced in this change?

Not individually in all cases. Some have specific tracking items. But we have a line item in the release plan to review all TODOs again for the next phase of the release. After that we can create items for whatever is not addressed.

**Pull request merged**